### PR TITLE
Streaming: improve JSDocs for DataSourceAPI streaming support

### PR DESCRIPTION
**Pull request by ryantxu**
After discussing streaming support with @aocenas, it is clear we are missing good docs somewhere :)

This is an attempt to add better JSDocs.  We likely need a blog post and more, but I hope to do that when we have a more coherent streaming solution.  ie, same thing works in grafana and explore

### CODE REVIEW

File: packages/grafana-ui/src/types/datasource.ts [hunk 1]
```diff
@@ -295,7 +313,7 @@ export interface ExploreStartPageProps {
  */
 export type LegacyResponseData = TimeSeries | TableData | any;
 
-export type DataQueryResponseData = DataFrameDTO | LegacyResponseData;
+export type DataQueryResponseData = DataFrame | DataFrameDTO | LegacyResponseData;
```
Comment (ryantxu) [inline packages/grafana-ui/src/types/datasource.ts]: unrelated... but I added it while I'm looking

Comment (hugohaggmark) [reply packages/grafana-ui/src/types/datasource.ts]: Shouldn't it be DataFrameHelper as well?

File: packages/grafana-ui/src/types/datasource.ts [hunk 2]
```diff
@@ -306,7 +326,25 @@ export interface DataStreamState {
...
+   *   data:[A]
+   * }
+   *
+   * Events:
+   * 1. {key:K1, data:[B1]}    >> PanelData: [A,B1]
+   * 2. {key:K2, data:[C2,D2]} >> PanelData: [A,B1,C2,D2]
+   * 3. {key:K1, data:[B3]}    >> PanelData: [A,B3,C2,D2]
+   * 4. {key:K2, data:[C4]}    >> PanelData: [A,B3,C4]
+   *
+   * NOTE: that PanelData will report a `Done` state until all
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/types/datasource.ts]:
```suggestion
   * NOTE: that PanelData will not report a `Done` state until all
```

File: packages/grafana-ui/src/types/datasource.ts [hunk 3]
```diff
@@ -318,7 +356,9 @@ export interface DataStreamState {
   request: DataQueryRequest;
 
   /**
-   * Data may not be known yet
+   * The streaming events return entire DataFrames.  The DataSource
+   * sending the events is responcible for truncating any growing lists
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/types/datasource.ts]:
```suggestion
   * sending the events is responsible for truncating any growing lists
```

**Review by hugohaggmark — APPROVED** [review]
LGTM, made some minor comments.

File: packages/grafana-ui/src/types/datasource.ts [hunk 4]
```diff
@@ -306,7 +326,25 @@ export interface DataStreamState {
...
+   * }
+   *
+   * Events:
+   * 1. {key:K1, data:[B1]}    >> PanelData: [A,B1]
+   * 2. {key:K2, data:[C2,D2]} >> PanelData: [A,B1,C2,D2]
+   * 3. {key:K1, data:[B3]}    >> PanelData: [A,B3,C2,D2]
+   * 4. {key:K2, data:[C4]}    >> PanelData: [A,B3,C4]
+   *
+   * NOTE: that PanelData will report a `Done` state until all
+   * unique keys have returned with either `Error` or `Done` state.
```
Comment (aocenas) [inline packages/grafana-ui/src/types/datasource.ts]: Whoa, are you saying that this is basically a stream of streams that you have to manually check by the key? Honestly did not see any code like that so far so not sure what to make from this. It seems like all of this is meant not to be used directly but only through `PanelQueryRunner` or something but that isn't the case right now so all these assumptions could be broken somewhere.

File: packages/grafana-ui/src/types/datasource.ts [hunk 5]
```diff
@@ -184,9 +187,26 @@ export abstract class DataSourceApi<
...
+   * Are you reading these docs aiming to execute a query?
+   * +-> If Yes, then consider using panelQueryRunner/State instead.  see:
+   *  * {@link https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryRunner.ts PanelQueryRunner.ts}
+   *  * {@link https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryState.ts PanelQueryState.ts}
+   *
+   * If you are implementing a simple request-response query,
+   * then you can ignore the `observer` entirely.
+   *
+   * When streaming behavior is required, the Promise can return at any time
+   * with empty or partial data in the response and optionally a state.
```
Comment (aocenas) [inline packages/grafana-ui/src/types/datasource.ts]: This I think is broken here https://github.com/grafana/grafana/pull/16676/files#r316658833. even though that is older PR I feel like these kind of errors are too easy to make here.

File: packages/grafana-ui/src/types/datasource.ts [hunk 6]
```diff
@@ -318,7 +356,9 @@ export interface DataStreamState {
   request: DataQueryRequest;
 
   /**
-   * Data may not be known yet
+   * The streaming events return entire DataFrames.  The DataSource
+   * sending the events is responcible for truncating any growing lists
+   * most likely to the requested `maxDataPoints`
```
Comment (aocenas) [inline packages/grafana-ui/src/types/datasource.ts]: No idea here why this is array of dataframes and not just a dataframe. At least on Explore side this seems to be flattened anyway.

Comment (ryantxu) [reply packages/grafana-ui/src/types/datasource.ts]: a query often returns many DataFrames.  and again remember, the existing query request contains a *list* of queries, not a single query

File: packages/grafana-ui/src/types/datasource.ts (hunk 4 above)
Comment (ryantxu) [reply packages/grafana-ui/src/types/datasource.ts]: > assumptions could be broken
well obviously!

> stream of streams
remember that the query argument is not a single query.  it is a request with a query[] (targets),  given that each query can return an array of data, that is how this mess gets made

File: packages/grafana-ui/src/types/datasource.ts (hunk 1 above)
Comment (ryantxu) [reply packages/grafana-ui/src/types/datasource.ts]: no, `DataFrameHelper` implements `DataFrame` -- it is just a helper ;)

**Comment by ryantxu** [comment]
@aocenas I added these docs as it defines the current behavior, I'll take a stab at an updated API that incorporates your feedback

File: packages/grafana-ui/src/types/datasource.ts (hunk 6 above)
Comment (torkelo) [reply packages/grafana-ui/src/types/datasource.ts]: This cannot be flattened in Explore, as a time series is presented as a single DataFrame (with time and value field), and Explore needs to show more than one time series :) A single query can return many series (explore also supports multiple queries which of course can return multiple series (ie DataFrames)

**Pull request merged**